### PR TITLE
Topic/mutable mk csr

### LINKED ISSUE #220: mkCSR seems to drop all 0 rows.
```
λ> mkCSR  [((0,999),1.0),((1,998),4.0) ]
CSR {csrVals = [1.0,4.0], csrCols = [1000,999], csrRows = [1,2,3], csrNRows = 2, csrNCols = 1000}
λ> mkCSR  [((1,999),1.0),((2,998),4.0) ]
CSR {csrVals = [1.0,4.0], csrCols = [1000,999], csrRows = [1,2,3], csrNRows = 2, csrNCols = 1000}
```
I would have expected the result from the second operation to be
```
CSR {csrVals = [1.0,4.0], csrCols = [1000,999], csrRows = [1,1,2,3], csrNRows = 3, csrNCols = 1000}
```

### LINKED ISSUE #320: mkSparse drops all-zero rows
Here is an example: 

Consider the (dense) matrix: 
`fromLists [[1,0],[0,1],[0,0],[1,1],[0,0],[0,0],[1,0],[0,0],[1,0]]`

If I create a sparse matrix using the following `AssocMatrix` representing the above matrix: 
`[((0,0), 1.0), ((1,1), 1.0), ((3,0), 1.0), ((3,1), 1.0), ((6,0), 1.0), ((8,1), 1.0)]`

`mkSparse` evaluates to: 
`SparseR {gmCSR = CSR {csrVals = [1.0,1.0,1.0,1.0,1.0,1.0], csrCols = [1,2,1,2,1,2], csrRows = [1,2,3,5,6,7], csrNRows = 5, csrNCols = 2}, nRows = 5, nCols = 2}`

whereas it should really evaluate to: 
`SparseR (CSR (vector [1.0,1.0,1.0,1.0,1.0,1.0]) ((fromList [1,2,1,2,1,2]) :: Vector CInt) ((fromList [1,2,3,3,5,5,5,6,6,7]) :: Vector CInt) 9 2) 9 2)`

Edit:
Maybe I will fix `mkCSR`, if I have time. In the meantime, I have bodged it by throwing a column vector of zeros into the `AssocMatrix` but, obviously, that's not very clean.

### PR DESCRIPTION
**Pull request by HuwCampbell**
Here's an alternative which permits streaming. It's not super polished yet or exposed; but if you agree that it's worthwhile I'll polish it up.

Fixes #220 
Fixes #320

### CODE REVIEW

**Comment by HuwCampbell** [comment]
I did some benchmarking, it looks like this version can be 5 to 10 times faster than the original version. Small medium and large are 100, 10k, and 1M entries in a diagonal matrix.


```
benchmarking mkCSR/small
time                 4.283 μs   (4.125 μs .. 4.556 μs)
                     0.990 R²   (0.979 R² .. 1.000 R²)
mean                 4.175 μs   (4.137 μs .. 4.315 μs)
std dev              212.5 ns   (62.74 ns .. 464.8 ns)
variance introduced by outliers: 64% (severely inflated)

benchmarking mkCSR/medium
time                 584.9 μs   (574.8 μs .. 594.7 μs)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 589.6 μs   (584.6 μs .. 593.5 μs)
std dev              14.87 μs   (12.46 μs .. 18.52 μs)
variance introduced by outliers: 16% (moderately inflated)

benchmarking mkCSR/large
time                 164.4 ms   (141.0 ms .. 191.3 ms)
                     0.985 R²   (0.968 R² .. 1.000 R²)
mean                 177.0 ms   (170.1 ms .. 183.1 ms)
std dev              9.480 ms   (7.195 ms .. 12.86 ms)
variance introduced by outliers: 13% (moderately inflated)

benchmarking original/small
time                 23.86 μs   (23.51 μs .. 24.18 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 23.79 μs   (23.62 μs .. 24.02 μs)
std dev              613.5 ns   (482.6 ns .. 783.5 ns)
variance introduced by outliers: 26% (moderately inflated)

benchmarking original/medium
time                 10.34 ms   (9.243 ms .. 11.33 ms)
                     0.934 R²   (0.893 R² .. 0.965 R²)
mean                 10.18 ms   (9.335 ms .. 11.11 ms)
std dev              2.369 ms   (1.794 ms .. 3.439 ms)
variance introduced by outliers: 86% (severely inflated)

benchmarking original/large
time                 1.131 s    (1.103 s .. 1.165 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.222 s    (1.181 s .. 1.269 s)
std dev              54.45 ms   (21.30 ms .. 69.86 ms)
variance introduced by outliers: 19% (moderately inflated)
```

**Comment by idontgetoutmuch** [comment]
Hi @HuwCampbell, Does this still need a review? Is there anything else that I can help on?

**Comment by HuwCampbell** [comment]
Yep, a review would be good. I think it's ready.

**Comment by idontgetoutmuch** [comment]
@HuwCampbell how are you running those benchmarks?

**Comment by HuwCampbell** [comment]
I just made up a little benchmark suite. IIRC, I had to add bangs to the constructor for CSR for it to actually measure the old version.

```haskell
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}
module Main where

import Criterion
import Criterion.Main

import Numeric.LinearAlgebra.Devel
import qualified Data.Vector.Storable as V


main :: IO ()
main = do
  let
    compile n = fmap (\i -> ((i, i), 10)) [0 .. n]
    small     = compile 100
    medium    = compile 10000
    large     = compile 1000000

  defaultMain [
      bgroup "mkCSR" [
        bench "small"  $ whnf mkCSR small
      , bench "medium" $ whnf mkCSR medium
      , bench "large"  $ whnf mkCSR large
      ]
  ]
```

**Comment by idontgetoutmuch** [comment]
I made a few minor suggestions

**Comment by HuwCampbell** [comment]
> I made a few minor suggestions

Nothing's shown up, you might need to complete the review for them to show.

**Comment by idontgetoutmuch** [comment]
Ah sorry I created a PR: https://github.com/HuwCampbell/hmatrix/pull/1

File: packages/base/src/Internal/Sparse.hs [hunk 1]
```diff
@@ -49,22 +52,77 @@ data CSC = CSC
...
+    (?) = flip
+    sfi = succ . fi
+    begin = do
+      mv <- M.unsafeNew 64
+      mr <- M.unsafeNew 64
+      mc <- M.unsafeNew 64
+      return (mv, mr, mc, 0, 0, 0, -1)
+
+    next (!mv, !mr, !mc, !idxVC, !idxR, !maxC, !curRow) ((r,c),d) = do
+      r < curRow ~!~ printf "impureCSR: row %i specified after %i" r curRow
```
Comment (idontgetoutmuch) [inline packages/base/src/Internal/Sparse.hs:90]: I suggest removing the definition for ~!~ I don't think it adds much

```
      when (r < curRow) $ error (printf "impureCSR: row %i specified after %i" r curRow)
```

File: packages/base/src/Internal/Sparse.hs [hunk 2]
```diff
@@ -49,22 +52,77 @@ data CSC = CSC
...
+      mr' <-
+        if idxR >= lenR - 1 then
+          M.unsafeGrow mr lenR
+        else
+          return mr
+
+      M.unsafeWrite mc' idxVC (sfi c)
+      M.unsafeWrite mv' idxVC d
+
+      idxR' <- foldM ? idxR ? [1 .. (r-curRow)] $ \idxR' _ -> do
```
Comment (idontgetoutmuch) [inline packages/base/src/Internal/Sparse.hs:113]: I suggest removing ? and just having

```
idxR' <- foldM (\idxR' _ -> M.unsafeWrite mr' idxR' (sfi idxVC) >> return (idxR' + 1))
               dxR [1 .. (r-curRow)]
```

**Comment by idontgetoutmuch** [comment]
@HuwCampbell these are just suggestions - the rest looks good to me - thanks very much - feel free to merge

File: packages/base/src/Internal/Sparse.hs (hunk 1 above)
Comment (HuwCampbell) [reply packages/base/src/Internal/Sparse.hs:90]: I agree.

File: packages/base/src/Internal/Sparse.hs (hunk 2 above)
Comment (HuwCampbell) [reply packages/base/src/Internal/Sparse.hs:113]: I'll trim to `succ idx <$ M.unsafeWrite mr' idxR' (sfi idxVC)`

**Comment by idontgetoutmuch** [comment]
@HuwCampbell if you make those changes then we can merge - thanks very much 😄

**Comment by HuwCampbell** [comment]
Done done. Cheers

**Pull request merged**